drivers/sensor/lis2dh/lis2dh.c

```diff
@@ -217,20 +217,25 @@ static int lis2dh_acc_odr_set(struct device *dev, u16_t freq)
 }
 #endif
 
-#ifdef CONFIG_LIS2DH_ACCEL_RANGE_RUNTIME
-static const union {
-	u32_t word_le32;
+/*
+ * Formula for the scale is:
+ *  ((SENSOR_G * 2 * (range_g)) / 65636LL)
+ */
+static const struct {
+	u16_t fs_scale[4];
 	u8_t fs_values[4];
-} lis2dh_acc_range_map = { .fs_values = {2, 4, 8, 16} };
+} lis2dh_acc_range_map = { .fs_scale = {583, 1168, 2335, 7007},
```

Comment:
Unfortunately those values do not come from formula in comment, they are slightly different. Either the comment or the code should be fixed.

Above values do not correspond to anything in the datasheet. I would rather use something from documentation such as Sensitivity (LIS2DH DS: 2.1. Mechanical characteristics):
```
#define LIS2DH_SCALE(sensitivity)               \                                                                                                                                                                                                                              
        ((SENSOR_G * (sensitivity)) >> 14)                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                               
static const u32_t lis2dh_reg_val_to_scale[] = {                                                                                                                                                                                                                               
        LIS2DH_SCALE(1),                                                                                                                                                                                                                                                       
        LIS2DH_SCALE(2),                                                                                                                                                                                                                                                       
        LIS2DH_SCALE(4),                                                                                                                                                                                                                                                       
        LIS2DH_SCALE(12),                                                                                                                                                                                                                                                      
};
```
and then assign `lis2dh->scale = lis2dh_reg_val_to_scale[fs];`.

Additionally I would take a chance to increase precision a bit, by chaning `lis2dh->scale` to 32 bit, shifting right sensitivity by 10 (instead of 14) and doing:
```
static void lis2dh_convert(s16_t raw_val, u32_t scale,
                           struct sensor_value *val)
{
        s32_t converted_val;

        /*
         * maximum converted value we can get is: max(raw_val) * max(scale)
         *      max(raw_val >> 4) = +/- 2^11
         *      max(scale) = 114921
         *      max(converted_val) = 235358208 which is less than 2^31
         */
        converted_val = (raw_val >> 4) * scale;
        val->val1 = converted_val / 1000000;
        val->val2 = converted_val % 1000000;
}
```

Reply:
> Above values do not correspond to anything in the datasheet. I would rather use something from documentation such as Sensitivity (LIS2DH DS: 2.1. Mechanical characteristics):

The values come directly from the LIS2DH_SCALE() formula, where I substituted  4, 8, 16, 32 with the values in the datasheet (i.e.  3.9, 7.82, 15.63, 46.9).

Anyway, to be honest, I don't like the way I implemented it. I did it too quickly.
So, I would take a look to your PR and we can merge that one. I think that you are doing an excellent job concerning the LIS2DH.

Reply:
> > Above values do not correspond to anything in the datasheet. I would rather use something from documentation such as Sensitivity (LIS2DH DS: 2.1. Mechanical characteristics):
> 
> The values come directly from the LIS2DH_SCALE() formula, where I substituted 4, 8, 16, 32 with the values in the datasheet (i.e. 3.9, 7.82, 15.63, 46.9).
> 
> Anyway, to be honest, I don't like the way I implemented it. I did it too quickly.
> So, I would take a look to your PR and we can merge that one. I think that you are doing an excellent job concerning the LIS2DH.

@mniestroj 
I understood the issue here. And it is a bad one, even if we can patch it.
As you probably know we are using the same driver for different devices that has same registers interface (i.e. whoami is same: 0x33). This devices currently are: LIS2DH, LIS3DH, LSM303DLHC, LIS2DH12, LSM303AGR (see drivers/sensor/lis2dh/Kconfig).

Now the issue is that, even if reg i/f is same, they have different mechanical sensors, so the scale is different. I was using the lsm303agr ones...

I will continue commenting on #20004

Reply:
@avisconti
So it looks like LSM303AGR is a special one from the whole set of supported sensors. I had a look into Linux drivers and they are using the same scale for all sensors that we support in this driver (`drivers/iio/accel/st_accel_core.c`):
```
#define ST_ACCEL_1_FS_AVL_2_GAIN		IIO_G_TO_M_S_2(1000)
#define ST_ACCEL_1_FS_AVL_4_GAIN		IIO_G_TO_M_S_2(2000)
#define ST_ACCEL_1_FS_AVL_8_GAIN		IIO_G_TO_M_S_2(4000)
#define ST_ACCEL_1_FS_AVL_16_GAIN		IIO_G_TO_M_S_2(12000)
```
So the question is whether we want to approach LSM303AGR sensor differently at this stage. The difference for LSM303AGR is <2.5% and the measurement error seems to be +/- 7%. So of course it would be nice to improve absolute measurements values, but the question is if we want to do it now.

Reply:
> @avisconti
> So it looks like LSM303AGR is a special one from the whole set of supported sensors. I had a look into Linux drivers and they are using the same scale for all sensors that we support in this driver (`drivers/iio/accel/st_accel_core.c`):
> 
> ```
> #define ST_ACCEL_1_FS_AVL_2_GAIN		IIO_G_TO_M_S_2(1000)
> #define ST_ACCEL_1_FS_AVL_4_GAIN		IIO_G_TO_M_S_2(2000)
> #define ST_ACCEL_1_FS_AVL_8_GAIN		IIO_G_TO_M_S_2(4000)
> #define ST_ACCEL_1_FS_AVL_16_GAIN		IIO_G_TO_M_S_2(12000)
> ```
> 
> So the question is whether we want to approach LSM303AGR sensor differently at this stage. The difference for LSM303AGR is <2.5% and the measurement error seems to be +/- 7%. So of course it would be nice to improve absolute measurements values, but the question is if we want to do it now.

@mniestroj 
Maybe we can add a separate set of macros to calculate the scale for lsm303agr, like LSM303AGR_SCALE(). What do you think?

Reply:
@avisconti 
Or something like this:
```
#if LSM303AGR
#define LIS2DH_SCALE_FACTOR (1172 / 1200)
#else
#define LIS2DH_SCALE_FACTOR 1
#endif

#define LIS2DH_SCALE(sensitivity)		\
	((SENSOR_G * (sensitivity) * LIS2DH_SCALE_FACTOR) >> 10)
```

Reply:
> @avisconti
> Or something like this:
> 
> ```
> #if LSM303AGR
> #define LIS2DH_SCALE_FACTOR (1172 / 1200)
> #else
> #define LIS2DH_SCALE_FACTOR 1
> #endif
> 
> #define LIS2DH_SCALE(sensitivity)		\
> 	((SENSOR_G * (sensitivity) * LIS2DH_SCALE_FACTOR) >> 10)
> ```
> 
> Just an idea, of course we would need to take integer arithmetic into account.

@mniestroj 
Two issues here:
1. I think it would be better to clarify that the scale factor is not for LIS2DH. So maybe we can use a better name.

2. Even if we find the correct arithmetic it is not stated in any datsheet, but would be only generated by comparison. So at least one comment is required.

But honestly I would prefer to have two different sets of macros.